File: panic.rs
#![allow(dead_code, clippy::redundant_closure, clippy::type_complexity)]

use std::panic;

fn main() {
    run();
    println!("Still alive.");
}

fn run() {
    let _hook_disable = PanicHookDisable::new();
    if let Err(err) = panic::catch_unwind(|| {
        for text in (1..=3).map(|x| process_text(x)) {
            println!("{}", text);
        }
    }) {
        let message = match err.downcast::<String>() {
            Ok(message) => *message,
            Err(err) => format!("{err:?}"),
        };
        println!("panic: {message}");
    }
}

fn process_text(id: i32) -> String {
    let text = retrieve_text(id).unwrap_or_else(|err| panic!("{err:?}"));
    text.chars().rev().collect()
}

static TEXTS: &[&str] = &["tar", "flow"];
// static TEXTS: &[&str] = &["tar", "flow", "🐻‍❄️❤️🦭"];

fn retrieve_text(id: i32) -> Result<String, Error> {
    match TEXTS.get((id - 1) as usize) {
        Some(text) => Ok(text.to_string()),
        _ => Err(Error::NotFound(format!("404 - Not found: {id}"))),
    }
}

#[derive(Debug)]
enum Error {
    NotFound(String),
}

struct PanicHookDisable {
    former: Option<Box<dyn Fn(&panic::PanicHookInfo) + Sync + Send + 'static>>,
}

impl PanicHookDisable {
    fn new() -> Self {
        let former = Some(panic::take_hook());
        panic::set_hook(Box::new(|_| {
            // std::process::exit(1);
        }));
        Self { former }
    }
}

impl Drop for PanicHookDisable {
    fn drop(&mut self) {
        panic::set_hook(self.former.take().unwrap());
    }
}

File: go/panic.go
package main

import "fmt"

func main() {
	run()
	fmt.Println("Still alive.")
}

func run() {
	defer func() {
		if r := recover(); r != nil {
			fmt.Println("panic:", r)
		}
	}()
	for i := 1; i <= 3; i++ {
		text := processText(i)
		fmt.Println(text)
	}
}

var texts = []string{"tar", "flow"}

func processText(id int) string {
	text, err := retrieveText(id)
	if err != nil {
		panic(err)
	}
	codes := []rune(text)
	reverseInPlace(codes)
	return string(codes)
}

func retrieveText(id int) (string, error) {
	if id >= 1 && id <= len(texts) {
		return texts[id-1], nil
	}
	// return "", fmt.Errorf("404 - Not found: %d", id)
	return "", NotFoundError{Message: fmt.Sprintf("404 - Not found: %d", id)}
}

func reverseInPlace[T any](values []T) {
	valuesLen := len(values)
	for index := 0; index < valuesLen/2; index += 1 {
		reverseIndex := valuesLen - index - 1
		values[index], values[reverseIndex] =
			values[reverseIndex], values[index]
	}
}

type NotFoundError struct {
	Message string
}

func (e NotFoundError) Error() string {
	return e.Message
}
